dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -5,6 +5,60 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
 
   # GET /users/auth/:provider/callback
   def all
+    if should_connect_provider?
+      connect_provider
+    else
+      login
+    end
+  end
+
+  # GET /users/auth/:provider/connect
+  def connect
+    return head(:bad_request) unless current_user&.migrated? && params[:provider].present?
+
+    session[:connect_provider] = 2.minutes.from_now
+    redirect_to omniauth_authorize_path(current_user, params[:provider])
+  end
+
+  # Call GET /users/auth/:provider/connect and the callback will trigger this code path
+  def connect_provider
+    return head(:bad_request) unless can_connect_provider? # TODO: (madelynkasula) better error here?
```

Comment:
I'm trying to figure out what the right behavior is here - if session[:connect_provider] has expired, then I'm guessing typically that means a user tried to connect an account, but didn't finish. So it's possible they might actually be trying to log in via a new provider, maybe? Since they'll be able to try again when we clear that variable, I think this works, we just may want to test and see what the failure case looks like to a user.

Reply:
yeah, i'm having a lot of trouble wrapping my head around this one. i'm hoping that once i write some tests it'll become a little clearer